### ISSUE TITLE
AuthMe/LuckPerms server stall

### ISSUE THREAD

**Issue by Eufranio** [AuthMe/AuthMeReloaded#1350]
The server just stalled once a (maybe) random player joined. Stall-report: https://gist.github.com/Eufranio/dd8feaf2df8f1c82731aa80a32506a8c
Config: https://gist.github.com/Eufranio/6738caa3dd617e00d82f14b94ec31bdc

Running:
  AuthMe-5.4-b1808
  Thermos-1.7.10-1614-b58

**Comment by sgdc3** [comment]
looks like a luckperms issue

**Comment by Eufranio** [comment]
@lucko?

**Comment by lucko** [comment]
@sgdc3 You're joining a future on the main thread.

This means the main thread is blocked, and pauses to wait for the future to complete. (not good)

You even commented this in the code....
https://github.com/AuthMe/AuthMeReloaded/blob/master/src/main/java/fr/xephi/authme/permission/handlers/LuckPermsHandler.java#L66

Don't see how you could possibly blame me, lol.

**Comment by sgdc3** [comment]
Whoops, missed that point, sorry xD

**Comment by sgdc3** [comment]
@lucko any way to set a timeout to the task?

**Comment by sgdc3** [comment]
Uh right, ComparableFuture.get(timeout, TimeUnit) ;)

**Comment by sgdc3** [comment]
@Eufranio Should have been fixed

**Comment by lucko** [comment]
Blocking the main server thread for 1 second is still absolutely not ok.

You will absolutely kill the TPS, and create 1 second long lag spikes (worst case) every time you retrieve an offline user.

**Comment by sgdc3** [comment]
> Also, it’s not a deadlock. You’re just waiting on a Future which can take a while to execute, as it does a lot of I/O, often with a remote database, and internal processing.

Probably a deadlock, because it took more than 30 seconds.

**Comment by sgdc3** [comment]
@lucko How could we check the permission of an offline user witout blocking the main thread, it's impossible if we don't cache it somehow.

**Comment by lucko** [comment]
Return a default value for the check, or actually perform the action which depends on the check asynchronously to start with.

**Comment by sgdc3** [comment]
@ljacqu any idea?

**Comment by ljacqu** [comment]
I'd like to forward the question to @lucko: when a player joins we do some permission checks (for VIP status and various restriction overriding permissions) based on the _name_ (not _Player_ object because it's not available in all joining events). I guess this falls into "offline player" check—how would you suggest we handle these?

**Comment by ljacqu** [comment]
Looking at the stacktrace, I see it failed while retrieving a player's groups.

@lucko Do we have a way to retrieve an online player's groups without Future?

**Comment by lucko** [comment]
Yes, just don't call the loadUser method.

**Comment by sgdc3** [comment]
@ljacqu that's why we have a hasPermission and hasPermissionOffline method

**Comment by sgdc3** [comment]
@ljacqu If we don't query the storage hasPermissionOffline will just return the dafault node values.

**Comment by lucko** [comment]
May I ask, when do you need to query offline users?

**Comment by sgdc3** [comment]
PlayerAsyncPreLoginEvent

**Comment by ljacqu** [comment]
@sgdc3 I know. Up until a week or two ago we queried online player permission check for join verifications. Maybe we should return to that. After all, we need the join process to be fast.

**Comment by lucko** [comment]
PreLogin event should be fine if you listen sometime after LOW. LP loads users then. 

Surely database purging is done async anyways? Same with admin commands which are likely to take a while to execute?

**Comment by Eufranio** [comment]
So?

**Comment by sgdc3** [comment]
@ljacqu i think we should make any permission check async for offline players

**Comment by ljacqu** [comment]
I think we shouldn't go over the offline player on join. If we don't have a known permission system the user cannot use permissions atm. This is not ideal

**Comment by Eufranio** [comment]
```
> [17:35:24 WARN]: java.util.concurrent.TimeoutException
> [17:35:24 WARN]: at java.util.concurrent.CompletableFuture.timedGet(CompletableFuture.java:1771)
> [17:35:24 WARN]: at java.util.concurrent.CompletableFuture.get(CompletableFuture.java:1915)
> [17:35:24 WARN]: at fr.xephi.authme.permission.handlers.LuckPermsHandler.getUser(LuckPermsHandler.java:75)
> [17:35:24 WARN]: at fr.xephi.authme.permission.handlers.LuckPermsHandler.getUser(LuckPermsHandler.java:66)
> [17:35:24 WARN]: at fr.xephi.authme.permission.handlers.LuckPermsHandler.getGroups(LuckPermsHandler.java:196)
> [17:35:24 WARN]: at fr.xephi.authme.permission.handlers.LuckPermsHandler.getGroups(LuckPermsHandler.java:33)
> [17:35:24 WARN]: at fr.xephi.authme.permission.PermissionsManager.getGroups(PermissionsManager.java:263)
> [17:35:24 WARN]: at fr.xephi.authme.data.limbo.LimboServiceHelper.createLimboPlayer(LimboServiceHelper.java:47)
> [17:35:24 WARN]: at fr.xephi.authme.data.limbo.LimboService.createLimboPlayer(LimboService.java:70)
> [17:35:24 WARN]: at fr.xephi.authme.process.join.AsynchronousJoin.lambda$processJoinSync$6(AsynchronousJoin.java:165)
> [17:35:24 WARN]: at org.bukkit.craftbukkit.v1_7_R4.scheduler.CraftTask.run(CraftTask.java:71)
> [17:35:24 WARN]: at org.bukkit.craftbukkit.v1_7_R4.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:346)
> [17:35:24 WARN]: at net.minecraft.server.MinecraftServer.func_71190_q(MinecraftServer.java:905)
> [17:35:24 WARN]: at net.minecraft.server.dedicated.DedicatedServer.func_71190_q(DedicatedServer.java:432)
> [17:35:24 WARN]: at net.minecraft.server.MinecraftServer.func_71217_p(MinecraftServer.java:841)
> [17:35:24 WARN]: at net.minecraft.server.MinecraftServer.run(MinecraftServer.java:693)
> [17:35:24 WARN]: at java.lang.Thread.run(Thread.java:748)
```

**Comment by lucko** [comment]
This is because your database is taking longer than 1 second to reply to the request, per this commit.

https://github.com/AuthMe/AuthMeReloaded/commit/fafbef6be72c12822fd05e2e9af7b5a5e4067f99

**Comment by Eufranio** [comment]
It is taking longer than 1 second because looks like AuthMe is lagging the server when players are joining...

**Comment by sgdc3** [comment]
@Eufranio no, it seems that your db response time is longer than 1 second, so authme has to halt the server and cause lag, we are going to move the permission checks to async, but please also check why your mysql db is so slow.

**Comment by ljacqu** [comment]
@sgdc3 What's the plan here? Given this and #1365 and the fact that a permission handler is now required I'd like to go back to our old approach

**Comment by sgdc3** [comment]
We should discuss that, maybe on discord

**Comment by Eufranio** [comment]
I'm almost certainly sure that the issue is not on my end, I have plenty of other applications/plugins using the same database server and they arent that slow. I also tried creating a SQLite database with the data of my MySQL one and testing with AuthMe, the SAME problem happens.

**Comment by krusic22** [comment]
I'm using LuckPerms (4x servers so 40 connection to DB) and AuthMe on the same DB (MariaDB) with a big connection limit and I'm having zero problems.
But you are using Thermos 1.7... So that might cause issues?

**Comment by Eufranio** [comment]
Is there a way for me to disable this integration at all, and use the old system? It was perfectly fine (for me), and I'd like to not have to rollback the AuthMe version and stick to an old build...

**Comment by sgdc3** [comment]
I could add an option to force using the vault hook instead

**Comment by sgdc3** [comment]
ok, that should work as a tempfix

**Comment by ljacqu** [comment]
> tempfix

@sgdc3 So what's the end game here? How will this be handled permanently?

**Comment by sgdc3** [comment]
@ljacqu i really don't know, luckperms should be handled fully async, but we are currently doing every permission check sync

**Comment by ljacqu** [comment]
@sgdc3 Can we revert the changes from using a `String name` back to a `Player` for those join verifications? I really dislike the implementation because we need to have a known perm system for permissions to work (counter-intuitive) and with the name we're basically doing an "offline player check" which isn't what we're doing (causing issues like this one)

**Comment by sgdc3** [comment]
@lucko That breaks permission purging

**Comment by ljacqu** [comment]
no, I'm speaking of the OnJoinVerifier. 

Purging happens in async so it's fine.

**Comment by sgdc3** [comment]
You mean ignoring the AsyncPreLoginEvent?

**Comment by sgdc3** [comment]
@ljacqu that should fix any issue

**Comment by ljacqu** [comment]
I mean undoing the part in 6c6fbaf12d282272789410cf04bc71857b15498e where `Player` gets changed to a `String name`. Is the performance improvement that good that it's worth all of this trouble? Because right now it's breaking stuff.

**Comment by sgdc3** [comment]
> Is the performance improvement that good that it's worth all of this trouble? Because right now it's breaking stuff.

In my opinion the performance improvement is worth the efford

**Comment by sgdc3** [comment]
> Because right now it's breaking stuff.

Just the LuckPerms integration, i didn't notice any other issue atm.

**Comment by ljacqu** [comment]
> we need to have a known perm system for permissions to work

**Comment by sgdc3** [comment]
LuckPerms is the only one that has that behaviour

**Comment by ljacqu** [comment]
Wrong https://github.com/AuthMe/AuthMeReloaded/blob/master/src/main/java/fr/xephi/authme/permission/PermissionsManager.java#L248
if no known perm handler, hasPermissionOffline won't do jack shit besides returning the default value, i.e. permission plugins unknown to us suddenly don't work with AuthMe.

**Comment by sgdc3** [comment]
@ljacqu ??? Any permission plugin that implements vault api correctly (fully sync compatible, not good for performance but consistent) is compatible.

**Comment by ljacqu** [comment]
Not every plugin hooks into vault

**Comment by sgdc3** [comment]
That is an huge performance improvement, it allows us to run every check async

**Comment by sgdc3** [comment]
We can make it configurable

**Comment by sgdc3** [comment]
Should have been fixed ;) @lucko we are now loading the user data on login in the AsyncPreLoginEvent, is that ok in your opinion?

**Comment by lucko** [comment]
Yep

**Comment by sgdc3** [comment]
Thank you
